### ISSUE TITLE
[Feature suggestion] Rotate graph layouts for nicer appearance

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#875]
I noticed a difference between Mathematica and igraph's layout method: Mathematica always rotates the layouts wide.  Here's a comparison of the same type of layout algorithm in both (Mathematica is on the top):

![image](https://cloud.githubusercontent.com/assets/1212871/10027966/2132e622-616a-11e5-9883-deb9b568d8fe.png)

This could be done through a simple principal component analysis and would be a nice feature for the high level interfaces. Just an idea :-) I implemented this for the Mathematica interface.

**Comment by ntamas** [comment]
The requirements for Python and R are a bit different I think. In Python, the plots are shown in a separate square-shaped window by default so there is no need to rotate the layout (although it would make sense to do another type of post-processing which would try to ensure that most edges are close to vertical or close to horizontal). However, when igraph is used in an IPython notebook, we are in a situation that is very similar to Mathematica's so it's probably makes sense to use a similar post-processing transformation.

At the moment I lack the spare time to implement such a post-processing transformation, but if anyone finds this interesting, feel free to comment and start working on this.

**Comment by szhorvat** [comment]
Actually what Mathematica does is that it breaks the graphs into connected components, lays out each one separately, rotates the layouts horizontally, then arranges the results in a nice way. There are multiple algorithms available for different arrangements, something similar to `igraph_layout_merge_dla`, I guess.  Check under Scope -> Packing Layout if you are interested: http://reference.wolfram.com/language/ref/GraphLayout.html

This is not a personal feature request, I don't need this. Just giving ideas :smile:

**Comment by szhorvat** [comment]
I now have principal component analysis-based (PCA-based) auto-rotation implemented in IGraph/M. It find the longest axis of the vertex coordinates and rotates it to be horizontal.

It does not work very well, and I think we should think about better rotation criteria. Here's an example where it works:

<img width="848" alt="image" src="https://user-images.githubusercontent.com/1212871/145408359-1da94724-e947-4232-9499-a8966a99555a.png">

And here is an example where it does not. This is Mathematica's own visualization of the nested line graph of a regular tree:

<img width="399" alt="image" src="https://user-images.githubusercontent.com/1212871/145408482-5ff23b9e-be23-4e41-92c2-390f41d36d55.png">

And here's igraph's Fruchterman-Reingold layout with PCA-based auto-rotation:

<img width="621" alt="image" src="https://user-images.githubusercontent.com/1212871/145408596-eb526cfc-d3fb-49cb-9d8c-8bc6c29ac22d.png">

Notice that in a triangle, there is no longest axis, so PCA cannot line up a size of the triangle with the horizontal axis.

We also see a failure of igraph's Fruchterman-Reingold (doesn't converge), but that's another topic.

**Comment by szhorvat** [comment]
@schochastics Please see what I wrote in chat. Would you be interested in prototyping something better than PCA for this?

**Comment by schochastics** [comment]
Something to brainstorm: This is a very simple optimization to try to make edges parallel to the axes. No fancy tools needed. Other constraints could be added to the objective function.

``` r
library(igraph)
# Rotate layout by angle (in radians)
rotate_layout <- function(layout, angle) {
  rotation_matrix <- matrix(c(cos(angle), -sin(angle), sin(angle), cos(angle)), ncol=2)
  layout %*% rotation_matrix
}

# Reflect layout over x or y axis
reflect_layout <- function(layout, axis = "none") {
  if (axis == "x") {
    layout[, 2] <- -layout[, 2]
  } else if (axis == "y") {
    layout[, 1] <- -layout[, 1]
  }
  layout
}

# Score: sum of squared deviation of edge angles from axis-aligned
edge_alignment_score <- function(layout, edges) {
  directions <- layout[edges[, 2], ] - layout[edges[, 1], ]
  angles <- atan2(directions[, 2], directions[, 1])

  angle_mod <- angles %% (pi / 2)
  deviation <- pmin(angle_mod, pi / 2 - angle_mod)
  sum(deviation^2)
}

optimize_layout <- function(layout, g) {
  edges <- as_edgelist(g, names = FALSE)
  axes <- c("none", "x", "y")
  best_score <- Inf
  best_layout <- layout

  for (axis in axes) {
    reflected <- reflect_layout(layout, axis)
    opt <- optimize(
      f = function(theta) {
        edge_alignment_score(rotate_layout(reflected, theta), edges)
      },
      interval = c(0, pi/2) # Enough due to symmetry
    )

    rotated <- rotate_layout(reflected, opt$minimum)
    score <- edge_alignment_score(rotated, edges)

    if (score < best_score) {
      best_score <- score
      best_layout <- rotated
    }
  }
  best_layout
}

# g <- make_tree(25, children = 2,mode="undirected")
g <- make_lattice(c(6,5), 2)
layout <- layout_with_fr(g)
layout_best <- optimize_layout(layout, g)
par(mfrow=(c(1, 2)))
plot(g, layout = layout)
plot(g, layout = layout_best)
```

![](https://i.imgur.com/yXARlfG.png)

<sup>Created on 2025-06-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by iosonofabio** [comment]
This looks interesting!

Once an implementation is merged, would it be ok for you guys if I add it to iplotx (Python)? That package is licensed MIT.

Thank you

**Comment by ntamas** [comment]
@iosonofabio We now have `igraph_layout_align()` in the C core and I believe @szhorvat is working on exposing this in Python.

**Comment by szhorvat** [comment]
Unfortunately I have to put exposing it in Python on hold as I need to complete some work on a deadline before leaving for holidays. @iosonofabio, you have experience exposing free functions (which we discussed this should be), so if you want to go ahead, that would be very helpful.
